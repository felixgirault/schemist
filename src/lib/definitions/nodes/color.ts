import type {NodeDef} from '$lib/stores/nodes';
import {defaultThemeColor} from '$lib/style';

export default {
	type: 'clr',
	label: 'Custom color',
	params: [
		{
			type: 'color',
			name: 'color',
			label: 'Color',
			default: defaultThemeColor
		}
	],
	samples: 'single',
	apply(_, {color}) {
		return color;
	}
} as NodeDef;
