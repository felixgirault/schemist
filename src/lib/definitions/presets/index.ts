import harmonies from './harmonies';
import materialDark from './materialDark';
import materialLight from './materialLight';
import materialNeutralDark from './materialNeutralDark';
import materialNeutralLight from './materialNeutralLight';
import materialTonesDark from './materialTonesDark';
import materialTonesLight from './materialTonesLight';
import neutral from './neutral';
import rainbow from './rainbow';
import scale from './scale';
import schemist from './schemist';
import semanticPairs from './semanticPairs';
import semantics from './semantics';
import splitComplementary from './splitComplementary';
import states from './states';
import tetrad from './tetrad';
import triad from './triad';

export default [
	states,
	splitComplementary,
	triad,
	tetrad,
	semantics,
	semanticPairs,
	scale,
	neutral,
	harmonies,
	rainbow,
	materialTonesLight,
	materialTonesDark,
	materialNeutralLight,
	materialNeutralDark,
	materialLight,
	materialDark,
	schemist
];
