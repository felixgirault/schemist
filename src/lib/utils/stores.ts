import type {Readable} from 'svelte/store';
import {derived} from 'svelte/store';
import {debounce} from './functions';
import type {DebounceTimeout} from './functions';

export const debouncedStore = <T>(
	store: Readable<T>,
	timeout: DebounceTimeout
) => {
	const debounced = debounce((value, set) => {
		set(value);
	}, timeout);

	let first = true;

	// `derived` relies on the callback signature to tell
	// if `set` is used, so we can't use `debounced` directly
	return derived<Readable<T>, T>(store, (value, set) => {
		if (first) {
			set(value);
			first = false;
		} else {
			debounced(value, set);
		}
	});
};
