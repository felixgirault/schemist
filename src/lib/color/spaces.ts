import type {
	Color,
	Definition,
	Hsl,
	Lch,
	Okhsl,
	Rgb
} from 'culori/fn';
import {
	clampChroma,
	formatCss,
	formatHex,
	modeHsl,
	modeLch,
	modeLrgb,
	modeOkhsl,
	modeOklch,
	modeRgb,
	parse,
	parseHex,
	useMode
} from 'culori/fn';
import {clamp, round} from '$lib/utils/math';

export interface SchemistColor extends Color {
	mode: 'schemist';
	h: number;
	s: number;
	l: number;
}

export interface Rgb255 extends Color {
	mode: 'rgb255';
	r: number;
	g: number;
	b: number;
}

export interface Hsl100 extends Color {
	mode: 'hsl100';
	h: number;
	s: number;
	l: number;
}

export interface Lch100 extends Color {
	mode: 'lch100';
	l: number;
	c: number;
	h: number;
}

export const toRgb255 = ({r, g, b, alpha}: Rgb): Rgb255 => ({
	mode: 'rgb255',
	r: r * 255,
	g: g * 255,
	b: b * 255,
	alpha
});

export const fromRgb255 = ({r, g, b, alpha}: Rgb255): Rgb => ({
	mode: 'rgb',
	r: r / 255,
	g: g / 255,
	b: b / 255,
	alpha
});

export const toHsl100 = ({h, s, l, alpha}: Hsl): Hsl100 => ({
	mode: 'hsl100',
	h,
	s: s * 100,
	l: l * 100,
	alpha
});

export const fromHsl100 = ({h, s, l, alpha}: Hsl100): Hsl => ({
	mode: 'hsl',
	h,
	s: s / 100,
	l: l / 100,
	alpha
});

export const toLch100 = ({l, c, h, alpha}: Lch): Lch100 => ({
	mode: 'lch100',
	l,
	c: (c / modeLch.ranges.c[1]) * 100,
	h,
	alpha
});

export const fromLch100 = ({l, c, h, alpha}: Lch100): Lch => ({
	mode: 'lch',
	l,
	c: (c / 100) * modeLch.ranges.c[1],
	h,
	alpha
});

const modeSchemist: Definition<SchemistColor> = {
	mode: 'schemist',
	channels: ['h', 's', 'l'],
	ranges: {
		h: [0, 360],
		s: [0, 100],
		l: [0, 100]
	},
	interpolate: {
		h: modeOklch.interpolate.h,
		s: modeOkhsl.interpolate.s,
		l: modeOkhsl.interpolate.l
	},
	toMode: {
		rgb: ({h, s, l, alpha}: SchemistColor) => {
			const hsl: Okhsl = {
				mode: 'okhsl',
				h,
				s: s / 100,
				l: l / 100
			};

			const lch = modeOklch.fromMode.oklab(
				modeOkhsl.toMode.oklab(hsl)
			);

			return modeOklch.toMode.rgb({...lch, h, alpha});
		}
	},
	fromMode: {
		rgb: (color: Rgb) => {
			const {h} = modeOklch.fromMode.rgb(color);
			const {s, l} = modeOkhsl.fromMode.rgb(color);

			return {
				mode: 'schemist',
				h,
				s: s * 100,
				l: l * 100,
				alpha: color.alpha
			};
		}
	}
};

export const rgbColor = useMode(modeRgb);
export const lrgbColor = useMode(modeLrgb);
export const hslColor = useMode(modeHsl);
export const lchColor = useMode(modeLch);
export const schemistColor = useMode(modeSchemist);

const converters = {
	rgb: rgbColor,
	lrgb: lrgbColor,
	hsl: hslColor,
	lch: lchColor,
	schemist: schemistColor
};

export const convertColor = (color, mode) =>
	converters[mode](color);

export const parseColor = parse;
export const isHexColor = (color: any) =>
	typeof color === 'string' && !!parseHex(color);

export const hexColor = formatHex;

export const cssColor = (
	color: Color,
	precision = 3
): string => {
	switch (color.mode) {
		case 'schemist':
			return 'alpha' in color
				? cssColor(rgbColor(clampChroma(color)))
				: hexColor(clampChroma(color));

		case 'rgb': {
			const {
				r = 0,
				g = 0,
				b = 0,
				alpha = 1
			} = toRgb255(color as Rgb);

			const rr = round(r, precision);
			const rg = round(g, precision);
			const rb = round(b, precision);
			const a = round(clamp(alpha, 1), precision);
			return a < 1
				? `rgba(${rr}, ${rg}, ${rb}, ${a})`
				: `rgb(${rr}, ${rg}, ${rb})`;
		}

		case 'hsl': {
			const {h = 0, s, l} = color as Hsl;

			return `hsl(${round(h, precision)} ${
				s === undefined
					? 'none'
					: round(s * 100, precision) + '%'
			} ${
				l === undefined
					? 'none'
					: round(l * 100, precision) + '%'
			})`;
		}

		case 'lch': {
			const {l, c, h = 0} = color as Lch;

			return `lch(${
				l === undefined ? 'none' : round(l, precision) + '%'
			} ${
				c === undefined ? 'none' : round(c, precision)
			} ${round(h, precision)})`;
		}

		default:
			return formatCss(color);
	}
};
