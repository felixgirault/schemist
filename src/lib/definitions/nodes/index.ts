import analogous from './analogous';
import color from './color';
import complementary from './complementary';
import contrasting from './contrasting';
import highlight from './highlight';
import hue from './hue';
import informative from './informative';
import lightness from './lightness';
import negative from './negative';
import positive from './positive';
import saturation from './saturation';
import shadow from './shadow';
import splitComplementaryLeft from './splitComplementaryLeft';
import splitComplementaryRight from './splitComplementaryRight';
import tetradLeft from './tetradLeft';
import tetradRight from './tetradRight';
import triadLeft from './triadLeft';
import triadRight from './triadRight';
import warning from './warning';

export default {
	[analogous.type]: analogous,
	[complementary.type]: complementary,
	[contrasting.type]: contrasting,
	[highlight.type]: highlight,
	[hue.type]: hue,
	[informative.type]: informative,
	[lightness.type]: lightness,
	[negative.type]: negative,
	[color.type]: color,
	[positive.type]: positive,
	[saturation.type]: saturation,
	[shadow.type]: shadow,
	[splitComplementaryLeft.type]: splitComplementaryLeft,
	[splitComplementaryRight.type]: splitComplementaryRight,
	[tetradLeft.type]: tetradLeft,
	[tetradRight.type]: tetradRight,
	[triadLeft.type]: triadLeft,
	[triadRight.type]: triadRight,
	[warning.type]: warning
};

export const defList = [
	analogous,
	complementary,
	splitComplementaryLeft,
	splitComplementaryRight,
	tetradLeft,
	tetradRight,
	triadLeft,
	triadRight,
	contrasting,
	informative,
	positive,
	warning,
	negative,
	highlight,
	shadow,
	lightness,
	saturation,
	hue,
	color
];
