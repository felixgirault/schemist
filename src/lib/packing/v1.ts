import {hexColor, schemistColor} from '$lib/color/spaces';
import defs from '$lib/definitions/nodes';
import type {PresetNode} from '$lib/presets';
import type {Args, Param} from '$lib/stores/nodes';

export const pack = (node: PresetNode): string => {
	const packed = [];
	const index = {};
	let i = 0;

	const indice = (key) => {
		if (!(key in index)) {
			index[key] = i++;
		}

		return index[key];
	};

	const packValue = (type: Param['type'], value: any) => {
		switch (type) {
			case 'boolean':
				return value ? 1 : 0;
			case 'range':
				return value;
			case 'color':
				return indice(hexColor(value).substring(1));
		}
	};

	const packArgs = (params: Param[], args: Args) =>
		params.flatMap(({type, name}) => [
			indice(name),
			packValue(type, args[name])
		]);

	const packNode = (node: PresetNode): number => {
		const typeIndex = indice(node.type);
		const tokenIndex = indice(node.token ?? '');
		const hiddenFlag = node.isHidden ? '!' : '';
		const args = node.args
			? packArgs(defs[node.type].params, node.args)
			: [];
		const children = node.children
			? node.children.map(packNode)
			: [];
		const i = packed.length;
		const data = [
			typeIndex,
			tokenIndex,
			args.length,
			...args,
			...children
		];

		const packedData = data
			.map((v) => v.toString(16))
			.map((v) =>
				v.length === 3
					? `y${v}`
					: v.length === 2
					? `x${v}`
					: v
			)
			.join('');

		packed.push(`${hiddenFlag}${packedData}`);
		return i;
	};

	const rootIndex = packNode(node).toString(16);

	return [
		rootIndex,
		packed.join(','),
		Object.keys(index).join(',')
	].join('@');
};

export const unpack = (data: string) => {
	const [rootIndex, inlinePacked, inlineIndex] =
		data.split('@');

	const packed = inlinePacked.split(',');
	const index = inlineIndex.split(',');

	const unpackValue = (type: Param['type'], value: number) => {
		switch (type) {
			case 'boolean':
				return !!value;
			case 'range':
				return value;
			case 'color':
				return schemistColor(`#${index[value]}`);
		}
	};

	const unpackArgs = (
		params: Param[],
		args: [number, number][]
	): Args => {
		const paramTypes = Object.fromEntries(
			params.map(({type, name}) => [name, type])
		);

		return Object.fromEntries(
			args.map(([k, v]) => {
				const name = index[k];
				return [name, unpackValue(paramTypes[name], v)];
			})
		);
	};

	const unpackData = (packedData: string) => {
		const matches = packedData.match(
			/y[0-9a-f]{3}|x[0-9a-f]{2}|[0-9a-f]/gi
		);

		const data = matches
			.map((v) =>
				v.startsWith('x') || v.startsWith('y')
					? v.substring(1)
					: v
			)
			.map((v) => parseInt(v, 16));

		const type = index[data.shift()];
		const token = index[data.shift()];
		const argCount = data.shift();
		const argEntries = [];

		for (let i = 0; i < argCount; i += 2) {
			argEntries.push([data.shift(), data.shift()]);
		}

		return {
			type,
			token,
			isHidden: packedData.startsWith('!'),
			args: unpackArgs(defs[type].params, argEntries),
			children: data
		};
	};

	const nodes = packed.map(unpackData);

	const inflateNode = ({
		type,
		token,
		isHidden,
		args,
		children
	}): PresetNode => ({
		type,
		token,
		isHidden,
		args,
		children: children.map((index) =>
			inflateNode(nodes[index])
		)
	});

	return inflateNode(nodes[parseInt(rootIndex, 16)]);
};
