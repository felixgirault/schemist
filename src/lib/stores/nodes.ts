import type {
	Readable,
	Unsubscriber,
	Writable
} from 'svelte/store';
import {derived, get, writable} from 'svelte/store';
import nearestColor from '$lib/color/nearest';
import {hexColor, schemistColor} from '$lib/color/spaces';
import type {SchemistColor} from '$lib/color/spaces';
import defs from '$lib/definitions/nodes';
import {withoutIndex} from '$lib/utils/arrays';
import {uid} from '$lib/utils/generators';
import {omit} from '$lib/utils/objects';
import {sentenceCase} from '$lib/utils/strings';

export type SampleType = 'single' | 'discrete' | 'continuous';

export type Param = {
	type: 'boolean' | 'range' | 'color';
	name: string;
	label: string;
	unit?: string;
	min?: number;
	max?: number;
	default: any;
};

export type Args = {
	[name: string]: any;
};

export type NodeDef = {
	type: string;
	label: string;
	argsLabel?: (args: Args) => string;
	description?: string;
	params: Param[];
	samples: SampleType;
	apply: (color: SchemistColor, args?: Args) => SchemistColor;
};

export type NodeOutput = [color: SchemistColor, name: string];
export type NodeOutputRecord = Record<Node['id'], NodeOutput>;

export type Node = {
	id: string;
	def: NodeDef;
	token: Writable<string>;
	isHidden: Writable<boolean>;
	input: Writable<NodeOutput>;
	args: Writable<Args>;
	output: Readable<NodeOutput>;
	children: Readable<Node[]>;
	parent?: Node;
	disconnectFromParent?: Unsubscriber;
	addChild: (node: Node) => void;
	removeChild: (node: Node) => void;
	connect: (outputs: Writable<NodeOutputRecord>) => void;
	destroy: () => void;
};

export const defaultArgs = (def) =>
	Object.fromEntries(
		def.params.map(({name, default: d}) => [name, d])
	);

export const generateName = (
	token: string,
	color: SchemistColor,
	parentColorName?: string
) =>
	token
		? token.replace('$', parentColorName)
		: sentenceCase(nearestColor(hexColor(color)));

// This whole thing is quite convoluted (shall I say shitty?)
// but at least it makes the tree logic self contained.
export const createNode = (
	type: keyof typeof defs,
	isHidden = false,
	token = ''
): Node => {
	let self: Node;
	let connectedOutputs: Writable<NodeOutputRecord>;
	let disconnect: () => void;

	const id = `n${uid()}`;
	const def = defs[type];
	const isHiddenStore = writable(isHidden);
	const tokenStore = writable(token);
	const input: Writable<NodeOutput> = writable([
		schemistColor('#fff'),
		''
	]);

	const args = writable(
		Object.fromEntries(
			def.params.map((input) => [input.name, input.default])
		)
	);

	const output = derived(
		[tokenStore, input, args],
		([$token, $input, $args]) => {
			const color = def.apply($input[0], $args);
			const name = generateName($token, color, $input[1]);
			return [color, name] as NodeOutput;
		}
	);

	const children = writable([]);

	const removeChild = (child: Node) => {
		children.update((children) => {
			const i = children.findIndex((c) => c === child);

			if (i < 0) {
				return children;
			}

			child?.disconnectFromParent();
			child.parent = null;
			return withoutIndex(children, i);
		});
	};

	const removeFromParent = (node) => {
		node.parent?.removeChild(node);
	};

	const addChild = (child: Node) => {
		children.update(($children) => {
			removeFromParent(child);

			child.parent = self;
			child.disconnectFromParent = output.subscribe(
				child.input.set
			);

			return $children.concat(child);
		});

		if (connectedOutputs) {
			child.connect(connectedOutputs);
		}
	};

	const destroy = () => {
		get(children).forEach((node) => node.destroy());
		removeFromParent(self);
		disconnect?.();
	};

	const connect = (outputs: Writable<NodeOutputRecord>) => {
		get(children).forEach((node) => node.connect(outputs));

		const disconnectOutput = output.subscribe((out) => {
			if (get(isHiddenStore)) {
				return;
			}

			outputs.update(($outputs) => ({
				...$outputs,
				[id]: out
			}));
		});

		const disconnectIsHidden = isHiddenStore.subscribe(
			(is) => {
				outputs.update(($outputs) =>
					is
						? omit($outputs, id)
						: {...$outputs, [id]: get(output)}
				);
			}
		);

		disconnect = () => {
			outputs.update(($outputs) => omit($outputs, id));
			disconnectOutput();
			disconnectIsHidden();
			disconnect = null;
		};

		connectedOutputs = outputs;
	};

	const {set, update, ...readOnlyChildren} = children;

	self = {
		id,
		def,
		isHidden: isHiddenStore,
		token: tokenStore,
		input,
		args,
		output,
		children: readOnlyChildren,
		addChild,
		removeChild,
		connect,
		destroy
	};

	return self;
};
