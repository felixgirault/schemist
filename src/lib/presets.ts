import {get} from 'svelte/store';
import type defs from '$lib/definitions/nodes';
import type {Args, Node, NodeDef} from '$lib/stores/nodes';
import {createNode} from '$lib/stores/nodes';

export type PresetNode = {
	type: keyof typeof defs;
	token?: string;
	args?: Args;
	children?: PresetNode[];
	isHidden?: boolean;
};

export type Preset = {
	label: string;
	description?: string;
	nodes: PresetNode[];
};

export const parseTree = (preset: PresetNode) => {
	const {type, isHidden, token, args, children} = preset;
	const node = createNode(type, isHidden, token);

	if (args) {
		node.args.update((current) => ({
			...current,
			...args
		}));
	}

	if (children) {
		children.forEach((childPreset) => {
			node.addChild(parseTree(childPreset));
		});
	}

	return node;
};

export const serializeTree = (node: Node) => {
	const token = get(node.token);
	const args = get(node.args);
	const children = get(node.children);
	const data: PresetNode = {
		type: node.def.type as keyof typeof defs,
		isHidden: get(node.isHidden)
	};

	if (token) {
		data.token = token;
	}

	if (Object.keys(args).length) {
		data.args = args;
	}

	if (children.length) {
		data.children = children.map(serializeTree);
	}

	return data;
};
