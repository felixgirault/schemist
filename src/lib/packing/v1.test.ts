import {expect, test} from 'vitest';
import {schemistColor} from '$lib/color/spaces';
import defs from '$lib/definitions/nodes';
import color from '$lib/definitions/nodes/color';
import presets from '$lib/definitions/presets';
import {defaultArgs} from '$lib/stores/nodes';
import {pack, unpack} from './v1';

test.each(Object.values(defs))('pack node $label', ({type}) => {
	const tree = {
		type,
		args: defaultArgs(defs[type])
	};

	const packed = pack(tree);
	const unpacked = unpack(packed);

	expect(packed).toMatchSnapshot();
	expect(unpacked).toMatchObject(tree);
});

test.each(Object.values(presets))(
	'pack preset $label',
	({nodes}) => {
		const tree = {
			type: color.type,
			args: {
				color: schemistColor('#f00')
			},
			children: nodes
		};

		const packed = pack(tree);
		const unpacked = unpack(packed);

		expect(packed).toMatchSnapshot();
		expect(unpacked).toMatchObject(tree);
	}
);
