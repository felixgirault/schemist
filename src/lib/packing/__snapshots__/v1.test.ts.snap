// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pack node Analogous 1`] = `"0@01221@anl,,steps"`;

exports[`pack node Complementary 1`] = `"0@010@cmp,"`;

exports[`pack node Contrasting 1`] = `"0@0122x50@ctr,,amount"`;

exports[`pack node Custom color 1`] = `"0@01223@clr,,color,ffc208"`;

exports[`pack node Highlight 1`] = `"0@0162a3f4y28a@hlt,,amount,shift,temp"`;

exports[`pack node Hue 1`] = `"0@0122xb4@h,,degrees"`;

exports[`pack node Informative 1`] = `"0@010@inf,"`;

exports[`pack node Left split complementary 1`] = `"0@010@scl,"`;

exports[`pack node Left tetrad 1`] = `"0@010@ttl,"`;

exports[`pack node Left triad 1`] = `"0@010@trl,"`;

exports[`pack node Lightness 1`] = `"0@0122x32@l,,amount"`;

exports[`pack node Negative 1`] = `"0@010@neg,"`;

exports[`pack node Positive 1`] = `"0@010@pos,"`;

exports[`pack node Right split complementary 1`] = `"0@010@scr,"`;

exports[`pack node Right tetrad 1`] = `"0@010@ttr,"`;

exports[`pack node Right triad 1`] = `"0@010@trr,"`;

exports[`pack node Saturation 1`] = `"0@0122x32@s,,amount"`;

exports[`pack node Shadow 1`] = `"0@0162a3f4y28a@sdw,,amount,shift,temp"`;

exports[`pack node Warning 1`] = `"0@010@wrn,"`;

exports[`pack preset Harmonies 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Material dark 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Material light 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Material neutral dark 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Material neutral light 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Material tones dark 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Material tones light 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Neutral 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Rainbow 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Scale 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Schemist theme 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Semantic color pairs 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Semantic colors 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Split complementary 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset States 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Tetrad 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;

exports[`pack preset Triad 1`] = `"14@670,890,45001,670,890,ab034,ef0,cd06,ef0,x10x1108,ef0,x12x130a,ef0,x14x150c,x19x1a2x18x62,x19x1b2x18x55,x19x1c2x18x32,x19x1d2x18f,x19x1e2x182,!x16x172x185efx10x11x12,012232579bdx13@clr,,color,ff0000,trl,Secondary,hlt,$ light,sdw,$ dark,trr,Tertiary,inf,Info,ctr,On $,pos,Success,wrn,Warning,neg,Error,s,Desaturated $,amount,l,Neutral white,Neutral light,Neutral,Neutral dark,Neutral black"`;
